src/clients/openai.rs

```diff
@@ -0,0 +1,423 @@
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum StopTokens {
+    Array(Vec<String>),
+    String(String),
+}
+
+// Chat completions API types
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ChatCompletionRequest {
+    /// ID of the model to use.
+    pub model: String,
+    /// A list of messages comprising the conversation so far.
+    pub messages: Vec<Message>,
+    #[serde(default)]
+    pub frequency_penalty: Option<f32>,
+    /// Modify the likelihood of specified tokens appearing in the completion.
+    #[serde(default)]
+    pub logit_bias: Option<HashMap<String, f32>>,
+    /// Whether to return log probabilities of the output tokens or not.
+    /// If true, returns the log probabilities of each output token returned in the content of message.
+    #[serde(default)]
+    pub logprobs: Option<bool>,
+    /// An integer between 0 and 20 specifying the number of most likely tokens to return
+    /// at each token position, each with an associated log probability.
+    /// logprobs must be set to true if this parameter is used.
+    #[serde(default)]
+    pub top_logprobs: Option<u32>,
+    /// The maximum number of tokens that can be generated in the chat completion.
+    #[serde(default)]
+    pub max_tokens: Option<u32>,
+    /// How many chat completion choices to generate for each input message.
+    #[serde(default)]
+    pub n: Option<u32>,
+    /// Positive values penalize new tokens based on whether they appear in the text so far,
+    /// increasing the model's likelihood to talk about new topics.
+    #[serde(default)]
+    pub presence_penalty: Option<f32>,
+    //#[serde(default)]
+    //pub response_format: Option<ResponseFormat>,
+    /// If specified, our system will make a best effort to sample deterministically,
+    /// such that repeated requests with the same seed and parameters should return the same result.
+    #[serde(default)]
+    pub seed: Option<u64>,
+    /// Up to 4 sequences where the API will stop generating further tokens.
+    #[serde(default)]
+    pub stop: Option<StopTokens>,
+    /// If set, partial message deltas will be sent, like in ChatGPT.
+    /// Tokens will be sent as data-only server-sent events as they become available,
+    /// with the stream terminated by a data: [DONE] message.
+    #[serde(default)]
+    pub stream: Option<bool>,
+    /// What sampling temperature to use, between 0 and 2.
+    /// Higher values like 0.8 will make the output more random,
+    /// while lower values like 0.2 will make it more focused and deterministic.
+    #[serde(default)]
+    pub temperature: Option<f32>,
+    /// An alternative to sampling with temperature, called nucleus sampling,
+    /// where the model considers the results of the tokens with top_p probability mass.
+    /// So 0.1 means only the tokens comprising the top 10% probability mass are considered.
+    #[serde(default)]
+    pub top_p: Option<f32>,
+
+    // Additional vllm params
```

Comment:
The reason it is included in this PR as the additional "provider" (issue references `ProviderType`, but took a different approach) for OpenAI is part of this PR, so it made since to include it. Is there any benefit of taking it out as it is already implemented (will create more work at this point)?

We have to maintain request/response types for all clients, so I don't really see an alternative to this. The API is pretty stable at this point as well.

Reply:
+1 on that.

We need to move towards a "passthrough" mechanism for this. Since we don't want to take API validation duty for vLLM or any other chat completion server.

Additionally, things change fast, there are already parameters deprecating and switching between vllm and openai and if we do strict validation, then we would end up being in the middle of their transition making our lives more difficult by having to constantly monitor and adhere to that transition.

Reply:
This code isn't doing any validation

Reply:
Per discussion, will update to support all modalities and add `max_completion_tokens` parameter so that it is fully compatible with the current spec.